### PR TITLE
Block editor: remove 4 useSelect in favour of context

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

#56847 has proven that the number of useSelect calls per block are definitely affecting performance. This PR removes another 4* of those separate store subscriptions, reuses the existing main block component subscription, and passes it down to slots and hooks through context (we already do that for clientId, name, etc.)

(*) Inspector slot, toolbar slot, block info slot, and one for hooks.

## Why?

See above.

## How?

Using context.

### CODE REVIEW

File: packages/block-editor/src/components/block-controls/hook.js [hunk 1]
```diff
@@ -6,14 +6,15 @@
  * Internal dependencies
  */
 import groups from './groups';
-import useDisplayBlockControls from '../use-display-block-controls';
+import { useBlockEditContext } from '../block-edit/context';
 
 export default function useBlockControlsFill( group, shareWithChildBlocks ) {
-	const { isDisplayed, isParentDisplayed } = useDisplayBlockControls();
-	if ( isDisplayed ) {
+	const { mayDisplayControls, mayDisplayParentControls } =
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-controls/hook.js]: So these are two new APIs?

Comment (ellatrix) [reply packages/block-editor/src/components/block-controls/hook.js]: This is internal context right?

Comment (ellatrix) [reply packages/block-editor/src/components/block-controls/hook.js]: Oh, I guess it's not... We could lock it with symbols I guess?

Comment (ellatrix) [reply packages/block-editor/src/components/block-controls/hook.js]: Fixed it 😊

**Review by youknowriad — APPROVED** [review]
Ok symbol as context keys as weird but if it's worth it in terms of performance, it seems fine. Maybe we should add a comment explaining the reasoning.

**Pull request merged**